### PR TITLE
Atom/galibzon/pad to n

### PR DESCRIPTION
**Pull request by galibzon**
Add [[pad_to(N)]] language feature 

This is version 1.7.34

"struct" definitions now support the [[pad_to(N)]] attribute.
This attribute, when used, must be declared after a member variable.
Each member variable can have one of these attributes after it.

The attribute is useful to pad the data with dummy variables until
guaranteeing that the offset of the next variable starts with the
desired alignment. The 'N' argument is a single integral literal
that must be a multiple of 4.

A companion class to the IntermediateRepresentation has been added:
StructPadToMutator. It contains all the logic and algorithms to
collect, during parsing, the variables that require padding
and apply the appropriate mutations (variable insertions) during
the IntermediateRepresentation::MiddleEnd().

### CODE REVIEW

File: src/AzslcIntermediateRepresentation.cpp [hunk 1]
```diff
@@ -92,8 +94,13 @@ namespace AZ::ShaderCompiler
                     m_metaData.m_outputFormatHint[rtIndex] = hint;
                 }
             }
+        } else if (attrInfo.m_attribute == "pad_to")
```
Comment (jeremyong-az) [inline src/AzslcIntermediateRepresentation.cpp]: Nit: `\n`

File: src/StructPadToMutator.cpp [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+            m_ir.ThrowAzslcIrException(IR_INVALID_PAD_TO_ARGUMENTS, attrInfo.m_lineNumber, errorMsg);
+        }
+        //Is the argument integral?
+        if (!holds_alternative<ConstNumericVal>(attrInfo.m_argList[0]))
+        {
+            string errorMsg("The [[pad_to(N)]] attribute only accepts one argument of integral type. A non integral argument was given.");
+            m_ir.ThrowAzslcIrException(IR_INVALID_PAD_TO_ARGUMENTS, attrInfo.m_lineNumber, errorMsg);
+        }
+        // Read the integral.
+        auto pad_to_value = static_cast<uint32_t>(ExtractValueAsInt64(get<ConstNumericVal>(attrInfo.m_argList[0]), 0));
```
Comment (jeremyong-az) [inline src/StructPadToMutator.cpp]: Might be worth adding `ExtractValueAsUint32` since this is probably a common case

File: src/StructPadToMutator.cpp [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+        if (!pad_to_value)
+        {
+            string errorMsg("Failed to read input integral to [[pad_to(N)]].");
+            m_ir.ThrowAzslcIrException(IR_INVALID_PAD_TO_ARGUMENTS, attrInfo.m_lineNumber, errorMsg);
+        }
+        // Must be a multiple of 16
```
Comment (jeremyong-az) [inline src/StructPadToMutator.cpp]: "multiple of 4"

File: src/StructPadToMutator.cpp [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+        }
+
+        auto& [curScopeId, curScopeKindInfo] = m_ir.GetCurrentScopeIdAndKind();
+        if (curScopeKindInfo.IsKindOneOf(Kind::Struct))
+        {
+            // We need to get the variable declared before this attribute.
+            auto varUid = m_ir.GetLastMemberVariable(curScopeId);
+            if (varUid.IsEmpty())
+            {
+                auto errorMsg = FormatString("The [[pad_to(N)]] attribute must be added after a member variable inside 'struct' or 'class'. The current scope '%.*s' doesn't have a declared variable yet.",
```
Comment (jeremyong-az) [inline src/StructPadToMutator.cpp]: Should we plan on permitting this in the future? (thinking of partial structs)

File: src/StructPadToMutator.cpp [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+                static_cast<int>(curScopeId.GetName().size()), curScopeId.GetName().data());
```
Comment (jeremyong-az) [inline src/StructPadToMutator.cpp]: Nit: spacing

File: src/StructPadToMutator.cpp [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+            {
+                const auto rows = exportedType.m_arithmeticInfo.m_rows;
+                const auto cols = exportedType.m_arithmeticInfo.m_cols;
+                const auto packAlignment = exportedType.m_arithmeticInfo.IsMatrix() ? Packing::Alignment::asMatrixStart : Packing::Alignment::asVectorStart;
+                startAt = offset = Packing::AlignOffset(layoutPacking, offset, packAlignment, rows, cols);
+            }
+
+            uint32_t totalArraySize = 1;
+            ArrayDimensions listOfArrayDim = varInfo.m_typeInfoExt.GetDimensions();
+            std::reverse(listOfArrayDim.m_dimensions.begin(), listOfArrayDim.m_dimensions.end());
```
Comment (jeremyong-az) [inline src/StructPadToMutator.cpp]: Nit: might be better to use a reverse iterator instead of allocating in the `ArrayDimensions` internal vector and copying

File: src/StructPadToMutator.cpp [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+
+        const auto* classInfo = m_ir.GetSymbolSubAs<ClassInfo>(exportedTypeId.m_name);
+        for (const auto& memberField : classInfo->GetMemberFields())
+        {
+            const auto currentStride = tempOffset;
+            CalculateMemberLayout(memberField, false, emitRowMajors, GetExtendedLayout(layoutPacking), tempOffset);
+            largestMemberSize = std::max(largestMemberSize, tempOffset - currentStride);
+        }
+
+        tempOffset = Packing::AlignStructToLargestMember(layoutPacking, tempOffset, largestMemberSize);
```
Comment (jeremyong-az) [inline src/StructPadToMutator.cpp]: Quick question: how exactly is the largest member impacting the overall alignment? Could you provide a quick example of what's happening here?

File: src/StructPadToMutator.cpp [hunk 8]
```diff
@@ -0,0 +1,479 @@
...
+            {
+                newVarInfo.m_typeInfoExt = ExtendedTypeInfo{ m_ir.m_sema.CreateTypeRefInfo(padType),
+                    {}, {{itemsCount}}, {}, Packing::MatrixMajor::Default };
+            }
+            newVarKind.GetSubRefAs<VarInfo>() = newVarInfo;
+            return newVarUid;
+        };
+
+        // The key idea is to add, at most, three variables. They will be added depending on keeping a 16-byte alignment from @startingOffset
+        // 1- The first variable will be added if @startingOffset is not 16-bytes aligned. It will be a float, float2 or float3.
```
Comment (jeremyong-az) [inline src/StructPadToMutator.cpp]: Out of curiosity, in what cases is this first variable needed? I imagine that between just the `float4`s plus any needed trailing padding bits this would be covered (the first `float4` if needed would align to a `float4` boundary, and the trailing bits would take care of the rest)

File: src/AzslcIntermediateRepresentation.cpp (hunk 1 above)
Comment (galibzon) [reply src/AzslcIntermediateRepresentation.cpp]: Fixed.

File: src/StructPadToMutator.cpp (hunk 2 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: Good idea. Added a template version called ExtractValueAs<T>()

File: src/StructPadToMutator.cpp (hunk 3 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: ooops. Fixed.

File: src/StructPadToMutator.cpp (hunk 4 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: When the time comes and the need arises, it will be "relatively" easy to add.

File: src/StructPadToMutator.cpp (hunk 5 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: Fixed.

File: src/StructPadToMutator.cpp (hunk 6 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: CalculateMemberLayout() & CalculateUserDefinedMemberLayout() are verbatim copies of
CodeReflection::BuildMemberLayout() & CodeReflection::BuildUserDefinedMemberLayout(), except that I removed the JSON code from it.

File: src/StructPadToMutator.cpp (hunk 7 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: Not sure, see my comment above. Those functions are varbatim copies of existing code that is used to calculate the SRG layout data, etc.

File: src/StructPadToMutator.cpp (hunk 8 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: For non-ConstantBuffer packing the float4 is not automatically aligned to 16 bytes.
Example:
```
struct MyStructA
{
         float m_data;
	 float4 m_arr[2];
};
```
For ConstantBuffer case you'll get these offsets:
```
;                   float m_data;                     ; Offset:    0
;                   float4 m_arr[2];                  ; Offset:   16
```
For StructuredBuffer case you'll get:
```
;           float m_data;                             ; Offset:    0
;           float4 m_arr[2];                          ; Offset:    4
```

File: src/StructPadToMutator.h [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ * 
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+#pragma once
+
+/*
+#include "AzslcUtils.h"
+*/
```
Comment (santorac) [inline src/StructPadToMutator.h]: Delete this?

File: src/StructPadToMutator.cpp [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+
+        auto& [curScopeId, curScopeKindInfo] = m_ir.GetCurrentScopeIdAndKind();
+        if (curScopeKindInfo.IsKindOneOf(Kind::Struct))
+        {
+            // We need to get the variable declared before this attribute.
+            auto varUid = m_ir.GetLastMemberVariable(curScopeId);
+            if (varUid.IsEmpty())
+            {
+                auto errorMsg = FormatString("The [[pad_to(N)]] attribute must be added after a member variable inside 'struct' or 'class'."
+                                             " The current scope '%.*s' doesn't have a declared variable yet.",
```
Comment (santorac) [inline src/StructPadToMutator.cpp]: Is this right?
```suggestion
                auto errorMsg = FormatString("The [[pad_to(N)]] attribute must be added after a member variable inside 'struct'."
                                             " The current scope '%.*s' doesn't have a declared variable yet.",
```

File: src/StructPadToMutator.cpp [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+            if (structItor == m_structsToPad.end())
+            {
+                m_structsToPad[curScopeId] = MapOfVarInfoUidToPadding();
+            }
+            MapOfVarInfoUidToPadding& varInfoUidToPadMap = m_structsToPad[curScopeId];
+            auto varInfoItor = varInfoUidToPadMap.find(varUid);
+            if (varInfoItor != varInfoUidToPadMap.end())
+            {
+                // It appears that there are two consecutive [[pad_to(N)]] attributes. This is an error.
+                auto errorMsg = string("Two consecutive [[pad_to(N)]] attributes are not allowed inside 'struct' or 'class'");
```
Comment (santorac) [inline src/StructPadToMutator.cpp]:
```suggestion
                auto errorMsg = string("Two consecutive [[pad_to(N)]] attributes are not allowed inside 'struct'");
```

File: src/StructPadToMutator.cpp [hunk 12]
```diff
@@ -0,0 +1,480 @@
...
+        if (visitedStructs.count(structUid))
+        {
+            return;
+        }
+        
+        const auto* classInfoPtr = m_ir.GetSymbolSubAs<ClassInfo>(structUid.m_name);
+        const auto listOfPairs = GetVariablesOfStructTypeThatRequirePadding(classInfoPtr);
+        for (const auto& [typeUid, varUid] : listOfPairs)
+        {
+            StructUidSortVisitFunction(typeUid, visitedStructs, sortedList);
```
Comment (santorac) [inline src/StructPadToMutator.cpp]: Is it possible to define structs that reference each other and still get to this point in the code? If so, you might hit infinite recursion and crash.

File: src/StructPadToMutator.cpp [hunk 13]
```diff
@@ -0,0 +1,480 @@
...
+        }
+        
+        visitedStructs.insert(structUid);
+        sortedList.push_back(structUid);
+    }
+
+
+    vector<pair<IdentifierUID, IdentifierUID>> StructPadToMutator::GetVariablesOfStructTypeThatRequirePadding(const ClassInfo* classInfo) const
```
Comment (santorac) [inline src/StructPadToMutator.cpp]: I'm curious why is padding supported for struct and not for class?

File: src/AzslcUtils.h [hunk 14]
```diff
@@ -230,6 +230,36 @@ namespace AZ::ShaderCompiler
...
+            {
+                return *defval;
+            }
+        }
+        else if (holds_alternative<float>(var))
+        {
+            // Casting from float has precision loss: https://onlinegdb.com/By0AJTUVE
+            auto floatVal = get<float>(var);
+            auto intVal = static_cast<int64_t>(floatVal);
+            PrintWarning(Warn::W3, none, "warning: Casting float ", floatVal, " to integer, will result in ", intVal);
```
Comment (jeremyong-az) [inline src/AzslcUtils.h:252]: This warning only applies if `T` is an integral type. We may need a small pass on this function to account for the various types `T` could vary over

File: src/StructPadToMutator.cpp [hunk 15]
```diff
@@ -0,0 +1,480 @@
...
+                {
+                    if (!m_ir.GetIdAndKindInfo(varInfo.GetTypeId().m_name))
+                    {
+                        continue;
+                    }
+
+                    // If array is a structure
+                    if (IsProductType(varClass))
+                    {
+                        size = CalculateUserDefinedMemberLayout(exportedType.m_typeId, emitRowMajor, layoutPacking, startAt = offset);
```
Comment (santorac) [inline src/StructPadToMutator.cpp]: Using "startAt = offset" inline could lead to confusion. I suggest putting this on its own line and pass startAt instead.

File: src/StructPadToMutator.cpp [hunk 16]
```diff
@@ -0,0 +1,480 @@
...
+
+            uint32_t totalArraySize = 1;
+            ArrayDimensions listOfArrayDim = varInfo.m_typeInfoExt.GetDimensions();
+            std::reverse(listOfArrayDim.m_dimensions.begin(), listOfArrayDim.m_dimensions.end());
+            for (const auto dim : varInfo.m_typeInfoExt.GetDimensions())
+            {
+                totalArraySize *= dim;
+            }
+
+            if (varInfo.m_typeInfoExt.IsArray() && !isArrayItr)
```
Comment (santorac) [inline src/StructPadToMutator.cpp]: What is "isArrayItr"? Could this variable be named better?

File: src/StructPadToMutator.cpp [hunk 17]
```diff
@@ -0,0 +1,480 @@
...
+                continue;
+            }
+
+            const auto bytesToAdd = alignedOffset - nextMemberOffset;
+            idx += InsertPaddingVariables(classInfo, structUid, idx+1, nextMemberOffset, bytesToAdd);
+            nextMemberOffset = alignedOffset;
+        }
+    }
+
+    uint32_t StructPadToMutator::CalculateMemberLayout(const IdentifierUID& memberId,
```
Comment (santorac) [inline src/StructPadToMutator.cpp]: It's pretty hard to follow what's going on in this function. You might consider adding more comments. I know you have some already, but it wasn't enough for me to easily understand.

File: src/StructPadToMutator.cpp [hunk 18]
```diff
@@ -0,0 +1,480 @@
...
+    {
+        // Alignment start
+        uint32_t tempOffset = startAt = Packing::AlignOffset(layoutPacking, startAt, Packing::Alignment::asStructStart, 0, 0);
+
+        uint32_t largestMemberSize = 0;
+
+        const auto* classInfo = m_ir.GetSymbolSubAs<ClassInfo>(exportedTypeId.m_name);
+        for (const auto& memberField : classInfo->GetMemberFields())
+        {
+            const auto currentStride = tempOffset;
```
Comment (santorac) [inline src/StructPadToMutator.cpp]: Nitpick: only use auto when it's saving long and ugly typenames. Here the type is "uint32_t" so just put that to make the code more readable.

File: src/StructPadToMutator.cpp (hunk 8 above)
Comment (santorac) [reply src/StructPadToMutator.cpp]: Consider including this info in code comments, not just PR comments.

**Review by santorac — COMMENTED** [review]
I didn't finish reviewing every file yet. I need to take a break and come back later.

File: src/StructPadToMutator.h (hunk 9 above)
Comment (galibzon) [reply src/StructPadToMutator.h]: Deleted. Thanks.

File: src/StructPadToMutator.cpp (hunk 10 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: Fixed. Thanks.

File: src/StructPadToMutator.cpp (hunk 12 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: I was thinking about that. Let me run some experiments. Maybe other earlier pieces of code prevent this from happening.

File: src/AzslcUtils.h (hunk 14 above)
Comment (galibzon) [reply src/AzslcUtils.h:252]: Oh yeah. Lots of things to check on T.

File: src/StructPadToMutator.cpp (hunk 11 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: Fixed.

File: src/StructPadToMutator.cpp (hunk 13 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: I removed support for 'class' types at the last minute. It was working, the algorithm even works for SRG.
Until I tried StructuredBuffer<class type> and DXC doesn't like that at all. The most common usage of [[pad_to]] is to facilitate the case of StructureBuffers and tightly packed structs that can also be used in ConstantBuffers.

If there's value in supporting 'class'es or SRGs the support is trivial to add.

File: src/StructPadToMutator.cpp (hunk 15 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: Good catch. That's copy/pasta from the original code that is used for SRG layout calculation. Fixed.

File: src/StructPadToMutator.cpp (hunk 16 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: As I mentioned in other comments: 
CalculateMemberLayout() & CalculateUserDefinedMemberLayout() are verbatim copies of
CodeReflection::BuildMemberLayout() & CodeReflection::BuildUserDefinedMemberLayout(), except that I removed the JSON code from it.

Lots of ugly code in there.

isArrayItr is flag that says the data type is iterating over an array definition.

I only removed the JSON related code from the original because it's code that's been working fine for the last two years.

File: src/StructPadToMutator.cpp (hunk 17 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: A neither do I understand much of it. Other than it's the same code used during SRG layout calculations.

File: src/StructPadToMutator.cpp (hunk 18 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: Agree, I hate auto (for the most part). It's easy on the coder, but really bad for the reviewer.
But the code base of AZSLc changed my soul.

File: src/AzslcUtils.h (hunk 14 above)
Comment (santorac) [reply src/AzslcUtils.h:252]: It looks like you only intend to use this for ints and floats. You could implement the function in cpp instead of h, and then explicitly instantiate just int and float. That will make it very clear what types are supported, while limiting the scope of error checking you need to do.

File: tests/Syntax/AsErrors/srg-pad-to.azsl [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+
+ShaderResourceGroupSemantic slot1
+{
+    FrequencyId = 1;
+};
+
+ShaderResourceGroup DemoSrg : slot1
+{
+    float m_red;
+    [[pad_to(16)]]
```
Comment (santorac) [inline tests/Syntax/AsErrors/srg-pad-to.azsl]: Maybe add more error cases to cover each of the many error checks  you put in the code: wrong args for pad_to, pad_to before any members, two pad_two's in a row, etc.

File: tests/Advanced/struct-pad-to.azsl [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+    [[pad_to(256)]]
+};
+
+ShaderResourceGroup DemoSrg : slot1
+{
+    MyStruct m_myStruct;
+    StructuredBuffer<MyStruct> m_mySB;
+};
+
+float4 PSMain() : SV_Target0
```
Comment (santorac) [inline tests/Advanced/struct-pad-to.azsl]: Does this test sufficiently cover all the variations of padding variables? You had the whole thee-variable "float1-3, float4[N], float1-3" thing going on in the code. I can't tell whether those cases are all covered here.

File: src/StructPadToMutator.cpp (hunk 12 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: Did some validations and AZSLc fails earlier when there are recursive usages of struct.

File: tests/Syntax/AsErrors/srg-pad-to.azsl (hunk 19 above)
Comment (galibzon) [reply tests/Syntax/AsErrors/srg-pad-to.azsl]: Done.

File: tests/Advanced/struct-pad-to.azsl (hunk 20 above)
Comment (galibzon) [reply tests/Advanced/struct-pad-to.azsl]: Added new test case that pads to 252, and exercises the case where three variables must be added to complete the padding.

File: src/StructPadToMutator.cpp (hunk 13 above)
Comment (jeremyong-az) [reply src/StructPadToMutator.cpp]: There's definitely value. One of the uses IIRC was adding padding to an SRG to ensure a value was an aligned load on Android

**Review by santorac — APPROVED** [review]

File: src/StructPadToMutator.cpp (hunk 13 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: Good to know, will add support for classes and SRGs.

Comment (galibzon) [reply src/StructPadToMutator.cpp]: [[pad_to(N)]] is now supported for classes and SRGs.

Comment (jeremyong-az) [reply src/StructPadToMutator.cpp]: Excellent!

File: src/StructPadToMutator.cpp (hunk 8 above)
Comment (galibzon) [reply src/StructPadToMutator.cpp]: Added the whole comment to the this function.

**Review by jeremyong-az — APPROVED** [review]
